File: packages/core_crypto/ios/core_crypto/Sources/core_crypto/NativeArgumentValidation.swift
import Flutter
import Foundation

enum NativeArgumentValidationError: Error {
  case invalid
}

let maxMnemonicUTF8Bytes = 512
let maxWordUTF8Bytes = 64
let maxKDFPasswordUTF8Bytes = 1024
let maxBackupPasswordUTF8Bytes = 4096
let maxSigningInputBytes = 1024 * 1024
let validBackupBlobSizes: Set<Int> = [60, 68, 76]
let supportedCoins: Set<String> = [
  "eth", "polygon", "base", "arbitrum", "avalanche", "bnb", "tron", "solana",
]
let validEntropySizes: Set<Int> = [16, 24, 32]
let validStoredWalletPayloadSizes: Set<Int> = [17, 25, 33, 61, 69, 77]

/// MethodChannel is a native trust boundary. Reject malformed direct calls
/// before authentication, Keychain, KDF, Wallet Core, or force casts.
func requireStringArgument(_ arguments: [String: Any], _ name: String) throws -> String {
  guard let value = arguments[name] as? String else {
    throw NativeArgumentValidationError.invalid
  }
  return value
}

func optionalStringArgument(_ arguments: [String: Any], _ name: String) throws -> String? {
  guard let raw = arguments[name], !(raw is NSNull) else { return nil }
  guard let value = raw as? String else { throw NativeArgumentValidationError.invalid }
  return value
}

func requireDataArgument(_ arguments: [String: Any], _ name: String) throws -> Data {
  guard let value = arguments[name] as? FlutterStandardTypedData else {
    throw NativeArgumentValidationError.invalid
  }
  return value.data
}

func requireIntArgument(_ arguments: [String: Any], _ name: String) throws -> Int {
  guard let value = arguments[name] as? Int else {
    throw NativeArgumentValidationError.invalid
  }
  return value
}

func optionalBoolArgument(
  _ arguments: [String: Any], _ name: String, default defaultValue: Bool
) throws -> Bool {
  guard let raw = arguments[name], !(raw is NSNull) else { return defaultValue }
  guard let value = raw as? Bool else { throw NativeArgumentValidationError.invalid }
  return value
}

func requireMnemonicStrength(_ arguments: [String: Any]) throws -> Int32 {
  let strength = try requireIntArgument(arguments, "strength")
  guard strength == 128 || strength == 192 || strength == 256 else {
    throw NativeArgumentValidationError.invalid
  }
  return Int32(strength)
}

func requireSuggestionLimit(_ arguments: [String: Any]) throws -> Int {
  let limit = try requireIntArgument(arguments, "limit")
  guard (1...20).contains(limit) else { throw NativeArgumentValidationError.invalid }
  return limit
}

private func requireUTF8Bounded(
  _ arguments: [String: Any], _ name: String, maxBytes: Int, allowBlank: Bool
) throws -> String {
  let value = try requireStringArgument(arguments, name)
  if !allowBlank && value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
    throw NativeArgumentValidationError.invalid
  }
  guard value.utf8.count <= maxBytes else { throw NativeArgumentValidationError.invalid }
  return value
}

func requireMnemonicText(_ arguments: [String: Any]) throws -> String {
  try requireUTF8Bounded(
    arguments, "mnemonic", maxBytes: maxMnemonicUTF8Bytes, allowBlank: false)
}

func requireWordText(_ arguments: [String: Any]) throws -> String {
  try requireUTF8Bounded(arguments, "word", maxBytes: maxWordUTF8Bytes, allowBlank: true)
}

func requireSuggestionPrefix(_ arguments: [String: Any]) throws -> String {
  try requireUTF8Bounded(arguments, "prefix", maxBytes: maxWordUTF8Bytes, allowBlank: true)
}

func optionalKDFPassword(_ arguments: [String: Any]) throws -> String? {
  guard let password = try optionalStringArgument(arguments, "kdfPassword") else { return nil }
  if password.isEmpty { return password }
  guard password.utf8.count <= maxKDFPasswordUTF8Bytes else {
    throw NativeArgumentValidationError.invalid
  }
  return password
}

func requireBackupPassword(_ arguments: [String: Any]) throws -> String {
  try requireUTF8Bounded(
    arguments, "password", maxBytes: maxBackupPasswordUTF8Bytes, allowBlank: false)
}

func requireSupportedCoin(_ arguments: [String: Any]) throws -> String {
  let coin = try requireStringArgument(arguments, "coin")
  guard supportedCoins.contains(coin) else { throw NativeArgumentValidationError.invalid }
  return coin
}

func requireSigningInput(_ arguments: [String: Any]) throws -> Data {
  let input = try requireDataArgument(arguments, "signingInput")
  guard !input.isEmpty, input.count <= maxSigningInputBytes else {
    throw NativeArgumentValidationError.invalid
  }
  return input
}

func requireBackupBlob(_ arguments: [String: Any]) throws -> Data {
  let blob = try requireDataArgument(arguments, "blob")
  guard validBackupBlobSizes.contains(blob.count) else {
    throw NativeArgumentValidationError.invalid
  }
  return blob
}

func requireStoredWalletFlag(_ flag: UInt8) throws -> UInt8 {
  guard flag == 0 || flag == 1 else { throw KeychainStore.StoreError.corrupted }
  return flag
}

func requireStoredWalletPayloadSize(_ payload: Data) throws -> Data {
  guard validStoredWalletPayloadSizes.contains(payload.count) else {
    throw KeychainStore.StoreError.corrupted
  }
  return payload
}

func requireEntropySize(_ entropy: Data) throws -> Data {
  guard validEntropySizes.contains(entropy.count) else {
    throw KeychainStore.StoreError.corrupted
  }
  return entropy
}

File: packages/core_crypto/android/src/main/kotlin/com/ktwallet/core_crypto/NativeArgumentValidation.kt
package com.ktwallet.core_crypto

internal class InvalidNativeArgumentException : Exception("invalid native argument")

internal const val MAX_MNEMONIC_UTF8_BYTES = 512
internal const val MAX_WORD_UTF8_BYTES = 64
internal const val MAX_KDF_PASSWORD_UTF8_BYTES = 1024
internal const val MAX_BACKUP_PASSWORD_UTF8_BYTES = 4096
internal const val MAX_SIGNING_INPUT_BYTES = 1024 * 1024
internal val VALID_BACKUP_BLOB_SIZES = setOf(60, 68, 76)
internal val SUPPORTED_COINS = setOf(
    "eth", "polygon", "base", "arbitrum", "avalanche", "bnb", "tron", "solana",
)
internal val VALID_ENTROPY_SIZES = setOf(16, 24, 32)
internal val PRIVATE_KEY_COPY_MODES = setOf("safe", "full")

internal fun requireExactArgumentKeys(arguments: Any?, expected: Set<String>) {
    val values = arguments as? Map<*, *> ?: throw InvalidNativeArgumentException()
    if (values.keys != expected) throw InvalidNativeArgumentException()
}

/**
 * MethodChannel is a native trust boundary. These helpers deliberately accept
 * [Any] so malformed direct calls are rejected before JNI, Keystore, KDF, or
 * authentication work instead of being coerced to defaults.
 */
internal fun requireNativeString(value: Any?): String =
    value as? String ?: throw InvalidNativeArgumentException()

internal fun optionalNativeString(value: Any?): String? = when (value) {
    null -> null
    is String -> value
    else -> throw InvalidNativeArgumentException()
}

internal fun requireNativeBytes(value: Any?): ByteArray =
    value as? ByteArray ?: throw InvalidNativeArgumentException()

internal fun requireNativeInt(value: Any?): Int =
    value as? Int ?: throw InvalidNativeArgumentException()

internal fun optionalNativeBoolean(value: Any?, default: Boolean): Boolean = when (value) {
    null -> default
    is Boolean -> value
    else -> throw InvalidNativeArgumentException()
}

internal fun requireMnemonicStrength(value: Any?): Int {
    val strength = requireNativeInt(value)
    if (strength !in setOf(128, 192, 256)) throw InvalidNativeArgumentException()
    return strength
}

internal fun requireSuggestionLimit(value: Any?): Int {
    val limit = requireNativeInt(value)
    if (limit !in 1..20) throw InvalidNativeArgumentException()
    return limit
}

private fun requireUtf8Bounded(value: Any?, maxBytes: Int, allowBlank: Boolean): String {
    val text = requireNativeString(value)
    if (!allowBlank && text.isBlank()) throw InvalidNativeArgumentException()
    val bytes = text.toByteArray(Charsets.UTF_8)
    return try {
        if (bytes.size > maxBytes) throw InvalidNativeArgumentException()
        text
    } finally {
        bytes.fill(0)
    }
}

internal fun requireMnemonicText(value: Any?): String =
    requireUtf8Bounded(value, MAX_MNEMONIC_UTF8_BYTES, allowBlank = false)

internal fun requireWordText(value: Any?): String =
    requireUtf8Bounded(value, MAX_WORD_UTF8_BYTES, allowBlank = true)

internal fun requireSuggestionPrefix(value: Any?): String =
    requireUtf8Bounded(value, MAX_WORD_UTF8_BYTES, allowBlank = true)

internal fun optionalKdfPassword(value: Any?): String? {
    val password = optionalNativeString(value) ?: return null
    if (password.isEmpty()) return password
    return requireUtf8Bounded(password, MAX_KDF_PASSWORD_UTF8_BYTES, allowBlank = false)
}

internal fun requireBackupPassword(value: Any?): String =
    requireUtf8Bounded(value, MAX_BACKUP_PASSWORD_UTF8_BYTES, allowBlank = false)

internal fun requireSupportedCoin(value: Any?): String {
    val coin = requireNativeString(value)
    if (coin !in SUPPORTED_COINS) throw InvalidNativeArgumentException()
    return coin
}

internal fun requirePrivateKeySessionId(value: Any?): String {
    val sessionId = requireNativeString(value)
    if (!Regex("^[A-Za-z0-9_-]{1,80}$").matches(sessionId)) {
        throw InvalidNativeArgumentException()
    }
    return sessionId
}

internal fun requirePrivateKeyCopyMode(value: Any?): String {
    val mode = requireNativeString(value)
    if (mode !in PRIVATE_KEY_COPY_MODES) throw InvalidNativeArgumentException()
    return mode
}

internal fun requireSigningInput(value: Any?): ByteArray {
    val input = requireNativeBytes(value)
    if (input.isEmpty() || input.size > MAX_SIGNING_INPUT_BYTES) {
        throw InvalidNativeArgumentException()
    }
    return input
}

internal fun requireBackupBlob(value: Any?): ByteArray {
    val blob = requireNativeBytes(value)
    if (blob.size !in VALID_BACKUP_BLOB_SIZES) throw InvalidNativeArgumentException()
    return blob
}

internal fun requireStoredWalletFlag(flag: Int): Int {
    if (flag != 0 && flag != 1) throw StoredWalletCorruptedException()
    return flag
}

internal fun requireEntropySize(entropy: ByteArray): ByteArray {
    if (entropy.size !in VALID_ENTROPY_SIZES) throw StoredWalletCorruptedException()
    return entropy
}
